geonode/security/utils.py

```diff
@@ -324,21 +274,348 @@ def get_user_visible_groups(user, include_public_invite: bool = False):
+            #       to the 'group' through the metadata settings.
+            # _resources = set([_r for _r in queryset.iterator()] + [_r for _r in group.resources()])
+            _resources = queryset.iterator()
+            for _r in _resources:
+                perm_spec = _r.get_all_level_info()
+                if perms:
+                    if "users" not in perm_spec:
+                        perm_spec["users"] = {}
+                    perm_spec["users"][user] = perms
+                # Let's the ResourceManager finally decide which are the correct security settings to apply
+                _r.set_permissions(perm_spec)
+
+        if not AdvancedSecurityWorkflowManager.is_auto_publishing_workflow():
+            if role and role == "manager":
+                _handle_perms(perms=VIEW_PERMISSIONS + DOWNLOAD_PERMISSIONS + BASIC_MANAGE_PERMISSIONS)
+            else:
+                _handle_perms(perms=VIEW_PERMISSIONS + DOWNLOAD_PERMISSIONS)
+
+    @staticmethod
+    def get_owner_permissions(uuid: str, /, instance=None, members: list = [], perm_spec: dict = {"users": {}, "groups": {}}) -> dict:
+        """compute a perm_spec to assing all admin permissions to the owner"""
+        _resource = instance or AdvancedSecurityWorkflowManager.get_instance(uuid)
+        _perm_spec = copy.deepcopy(perm_spec)
+
+        if _resource:
+            _resource = _resource.get_real_instance()
+            if _resource.polymorphic_ctype:
+                # Owner & Manager Admin Perms
+                admin_perms = VIEW_PERMISSIONS + ADMIN_PERMISSIONS
+                if _resource.polymorphic_ctype.name in DOWNLOADABLE_RESOURCES:
+                    admin_perms += DOWNLOAD_PERMISSIONS
+                for perm in admin_perms:
+                    if AdvancedSecurityWorkflowManager.is_auto_publishing_workflow():
+                        _prev_perm = _perm_spec["users"].get(_resource.owner, []) if "users" in _perm_spec else []
+                        _perm_spec["users"][_resource.owner] = set.union(perms_as_set(_prev_perm), perms_as_set(perm))
+                    elif perm not in {'change_resourcebase_permissions', 'publish_resourcebase'}:
+                        _prev_perm = _perm_spec["users"].get(_resource.owner, []) if "users" in _perm_spec else []
+                        _perm_spec["users"][_resource.owner] = set.union(perms_as_set(_prev_perm), perms_as_set(perm))
+                    if members:
+                        for user in members:
+                            _prev_perm = _perm_spec["users"].get(user, []) if "users" in _perm_spec else []
+                            _perm_spec["users"][user] = set.union(perms_as_set(_prev_perm), perms_as_set(perm))
+
+                # Set the GeoFence Owner Rule
+                if _resource.polymorphic_ctype.name == 'dataset':
+                    DATA_EDIT_PERMISSIONS = []
+                    if _resource.get_real_instance().subtype == 'vector':
```

Comment:
the `_resource` (line 427) is already the real instance, is not needed to do this operation again in the IF block

Reply:
done